libmisc/agetpass.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2022, Alejandro Colomar <alx@kernel.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the copyright holders or contributors may not be used to
+ *    endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <limits.h>
+#include <readpassphrase.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ident "$Id$"
+
+#include "prototypes.h"
+
+
+#if !defined(PASS_MAX)
+#define PASS_MAX  BUFSIZ
+#endif
+
+
+char *
+agetpass(const char *prompt)
+{
+	char    *pass;
+	size_t  len;
+
+	pass = malloc(PASS_MAX);
+	if (pass == NULL)
+		return NULL;
+
+	if (readpassphrase(prompt, pass, PASS_MAX, RPP_REQUIRE_TTY) == NULL)
+		goto fail;
+
+	len = strlen(pass);
+
+	if (len == 0)
+		return pass;
+
+	if (pass[len - 1] != '\n')
+		goto truncated;
+
+	pass[len - 1] = '\0';
+
+	return pass;
+
+truncated:
+	memzero(pass, PASS_MAX);
+fail:
+	free(pass);
+	return NULL;
```

Comment:
Can you reduce the number of goto definitions to one? It helps reading the code and the future maintenance tasks.

Moreover, if you are using them it would be nice to have just one return sentence (or maybe two if we can take into account the malloc error).

Reply:
Yes, we can reduce the `goto`s to one by calling `erase_pass()`.  Since I wrote the code for `agetpass()` before deciding to accept `NULL` in the cleanup, it wasn't possible, but now it is.  It'll also reduce code size a little bit.  I had this consideration before pushing the code, but in the end decided not to reduce the number of `goto`s, as I think maintainability will be better with two `goto`s, as the code won't depend on if `erase_pass(3)` accepts `NULL` or not (arguably this is unlikely to change), and also because the `truncated` label makes it clear the error condition without having to add a comment and without forcing the maintainer to guess it by the code.

Consider the change to accomodate to the suggestion:

```c
diff --git a/libmisc/agetpass.c b/libmisc/agetpass.c
index 619c9f36..e5be64a8 100644
--- a/libmisc/agetpass.c
+++ b/libmisc/agetpass.c
@@ -61,16 +61,14 @@ agetpass(const char *prompt)
                return pass;
 
        if (pass[len - 1] != '\n')
-               goto truncated;
+               goto fail;
 
        pass[len - 1] = '\0';
 
        return pass;
 
-truncated:
-       memzero(pass, PASS_MAX);
 fail:
-       free(pass);
+       erase_pass(pass);
        return NULL;
 }
 
```

I'd like to ask you to reconsider the suggestion, as that code is less readable to me (but of course this is subjective, and if you confirm that you like it more, I'll change it).  I strongly oppose to adding comments to the code, considering the `goto` label did a great job of distinguishing the error cases.

Regarding the one return sentence (not accounting the `malloc(3)` error), I'm not sure how to achieve it in a not-too-deeply-indented way.  Of course it can be done with lots of `if`-`else`, but I'd prefer to avoid that.  See <https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions>.

Reply:
BTW, I just asked to OpenBSD maintainers if `explicit_bzero(3)` is required after `readpassphrase(3)` failed.  It's not so clear as I first thought.

<https://marc.info/?l=openbsd-misc&m=166437328100790&w=2>

Reply:
I'd suggest the following:
```
char *
agetpass(const char *prompt)
{
	char    *pass;
	size_t  len;
	bool err = false;

	pass = malloc(PASS_MAX);
	if (pass == NULL)
		return NULL;

	if (readpassphrase(prompt, pass, PASS_MAX, RPP_REQUIRE_TTY) == NULL) {
		err = true;
		goto done;
	}

	len = strlen(pass);

	if (len == 0)
		goto done;

	if (pass[len - 1] != '\n') {
		memzero(pass, PASS_MAX);
		err = true;
		goto done;
	}

	pass[len - 1] = '\0';

done:
	if (err == true) {
	    free(pass);
	}

	return pass;
}
```

What do you think?

Reply:
That has a bug: we should return `NULL` on error, not a `free(3)`d pointer.

Reply:
I updated the code to have only one label.

Reply:
Thanks!